### PR TITLE
[WIP] Adding configuration to allow users to permit bearer token auth over HTTP instead of HTTPS via ClientOptions

### PR DESCRIPTION
**Pull request by schaabs**
This is an alternative implementation for the changes proposed in #8882, adding configuration to override the default behavior and allow bearer tokens to be sent over HTTP. In this implementation the configuration properties are added to `ClientOptions` rather than `TokenCredential`. This PR adds a new class `SecurityOptions` to `ClientOptions`. The `SecurityOptions` class introduces a property to allow bearer tokens to be sent over HTTP, namely `AllowInsecureConfidentialAuthenticationTransport`

I like the API surface of this implementation better then the implementation adding the configuration to `TokenCredential` as I feel this is a more natural place for this configuration to live. You are in essence configuring the client communication with the service, not the credential communication with the authority. This also lets me use the same credential with clients which are configured differently in this regard.

One drawback to this implementation is that it requires each client library to plumb the `ClientOptions` to the `BearerTokenAuthenticationPolicy` in order for the client to properly respect this configuration, so this change touches all client libraries vs. the `TokenCredential` implementation which only impacts Azure.Core and Azure.Identity.

Finally, I know the property `AllowInsecureConfidentialAuthenticationTransport` is quite a mouthful. I purposefully tried to create a daunting property name which was somewhat impenetrable to scare people off unless they specifically know they want to set this property. That being said I'd welcome any other suggestions for this.

### CODE REVIEW

File: sdk/core/Azure.Core/src/SecurityOptions.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+    /// <summary>
+    ///  The set of options that can be specified to influence how security constraints are enforced by the client.
+    /// </summary>
+    public class SecurityOptions
+    {
+        /// <summary>
+        /// Gets or sets a value determining if the client will allow confidential authentication schemes over non-TLS (non-HTTPS)
+        /// protected connections. WARNING: Setting this value can compromise the security of accounts used by your application.
+        /// </summary>
+        public bool AllowInsecureConfidentialAuthenticationTransport { get; set; } = false;
```
Comment (chlowell) [inline sdk/core/Azure.Core/src/SecurityOptions.cs]: `AllowInsecureAuthorizationTransport`?

File: sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs [hunk 2]
```diff
@@ -26,22 +28,46 @@ public class BearerTokenAuthenticationPolicy : HttpPipelinePolicy
...
+        {
+        }
+
+        /// <summary>
+        /// Creates a new instance of <see cref="BearerTokenAuthenticationPolicy"/> using provided token credential and scope to authenticate for.
+        /// </summary>
+        /// <param name="credential">The token credential to use for authentication.</param>
+        /// <param name="scope">The scope to authenticate for.</param>
+        /// <param name="options"><see cref="ClientOptions"/> that allow to configure the <see cref="BearerTokenAuthenticationPolicy"/>.</param>
+        public BearerTokenAuthenticationPolicy(TokenCredential credential, string scope, ClientOptions? options)
```
Comment (pakrym) [inline sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs]: I would take the bool here instead of the entire options class.

**Comment by pakrym** [comment]
I like this change even if it requires updating client to enable the workaround.

**Comment by jsquire** [comment]
@shaabs: Is this still active, or should we close this out as stale?

**Pull request closed without merging**